### PR TITLE
Theoretical fix for some crashes

### PR DESCRIPTION
**Pull request by zadjii-msft**
RE: 
* #15454
* MSFT:44725712 "WindowsTerminal.exe!NonClientIslandWindow::OnSize"
* MSFT:44754014 "NonClientIslandWindow::GetTotalNonClientExclusiveSize"

I think this should fix all of those, but I want to ship and verify live, since I can't repro this locally.

### CODE REVIEW

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 1]
```diff
@@ -82,6 +82,16 @@ void IslandWindow::Close()
...
+    //
+    // Specifically, this prevents a pending coroutine from being able to call
+    // something like ShowWindow, and have that come back on the IslandWindow
+    // message loop, where it'll end up asking XAML something that XAML is no
+    // longer able to answer.
+    SetWindowLongPtr(_window.get(), GWLP_USERDATA, (LONG_PTR)nullptr);
+
     if (_source)
     {
         _source.Close();
```
Comment (lhecker) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:97]: I just realized... When we do `winrt::detach_abi` above, it'll null out `_source` and so we won't call `Close()` anymore. Is that intentional or should we just remove `Close()`?

Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:97]: I mean, yea - but that's still only for Windows 10, and I'm blowing that bodge away in #15424 anyways

**Review by lhecker — APPROVED** [review]

File: src/cascadia/WindowsTerminal/IslandWindow.cpp [hunk 2]
```diff
@@ -82,6 +82,16 @@ void IslandWindow::Close()
...
 
+    // GH#15454: Unset the user data for the window. This will prevent future
+    // callbacks that come onto our window message loop from being sent to the
+    // IslandWindow (or other derived class's) implementation.
+    //
+    // Specifically, this prevents a pending coroutine from being able to call
+    // something like ShowWindow, and have that come back on the IslandWindow
+    // message loop, where it'll end up asking XAML something that XAML is no
+    // longer able to answer.
+    SetWindowLongPtr(_window.get(), GWLP_USERDATA, (LONG_PTR)nullptr);
```
Comment (DHowett) [inline src/cascadia/WindowsTerminal/IslandWindow.cpp:93]: I am not in love with the C-style cast, nor would audit mode be - does `reinterpret_cast<LONG_PTR>(nullptr)` work?

Comment (zadjii-msft) [reply src/cascadia/WindowsTerminal/IslandWindow.cpp:93]:
```suggestion
    SetWindowLongPtr(_window.get(), GWLP_USERDATA, 0);
```

**Review by DHowett — APPROVED** [review]

**Pull request merged**